rcl/src/rcl/node.c

```diff
@@ -284,6 +283,20 @@ rcl_node_init(
     // error message already set
     goto fail;
   }
+  // Initialize the node type cache hash map
+  ret = rcl_node_type_cache_init(node);
+  if (ret != RCL_RET_OK) {
+    // error message already set
+    goto fail;
+  }
```

Comment:
Sorry, one thing I just noticed here.  If the `get_type_description_service` service is disabled, we don't need the type cache at all, right?  That is, when it is disabled we can save memory by not allocating the cache and then not adding things to it.  So I'll suggest we move this inside of the `if (node->impl->options.enable_type_description_service) {`, and then have a similar `if` statement inside of `publisher_init`/`subscription_init`/`service_init`.

Reply:
One concern with this - if we allow the service to be turned on and off at runtime (something we discussed), then if the service was turned on after some sub/pub/srv creation, those types won't be available in its type cache.

Reply:
OK yeah, I'm back on board

Reply:
I've unfortunately got to turn this ship _back_ around.

The new design allows for the service to be enabled via Parameters, and initialization is therefore called from the instantiating client. However, in order for the type cache to have _all_ types the node provides - the type cache has to have been initialized before any interface is created (including the parameter topics). This has to be done in `rcl_node_init` because the rosout publisher is created in this function.

This comment originally banked on using NodeOptions for enabling, but if there is parameter-based configuration then there is no way to avoid always initializing the hashmap here.

I'll argue the overhead is negligible - no hashing calculation is done since the hashes are precomputed, and the data stored is `4 * sizeof(size_t)` per interface, which just doesn't add up to much